### ISSUE TITLE
A rant/suggestion about pos instances and substreams

### ISSUE THREAD

**Issue by dgelessus** [kaitai-io/kaitai_struct#709]
This is going to be a bit of a rant, but I'll try to keep it constructive. To be clear, this is a rant about *one particular aspect* of Kaitai Struct that I find a bit frustrating. I don't have anything against Kaitai Struct in general - quite the opposite :)

## The rant

Currently, Kaitai Struct's `pos` instances are heavily dependent on the mechanics of IO streams and substreams. When you write a `pos` instance, the `pos` offset is always interpreted relative to an IO stream - either one you explicitly specify using `io`, or the implicit default, which is the current structure's `_io`. I find this strange - most other parts of the KSY language try to be declarative and independent of any particular language or concrete parser implementation. `pos` instances are the complete opposite - to use them, you *have* to understand that the generated parser uses multiple IO (sub)streams, and when exactly it creates new ones or reuses existing ones.

The last part is what I find most frustrating in practice. Substreams are an important concept in advanced specs, but they are very difficult to understand, because they are never explicitly declared anywhere - you don't put `substream: true` on a field or type for example. Instead, substreams are only created implicitly, and the rules are not always obvious. When something has an explicit `size` (or `size-eos`), it gets a substream - except when it's a primitive type. Fixed-size types also don't count by default - for example, if you have a `chunk` type that is always 256 bytes long, then a field with `type: chunk` *doesn't* get a substream. If you want one, you have to explicitly set `size: sizeof<chunk>` on the field, which looks redundant if you don't know the mechanics of substreams.

I'm assuming that the current mechanics of `pos` instances and substreams come from how the KSC-generated parsers are currently implemented. This of course simplifies code generation, but I could see this becoming a problem in the long term. By exposing the implementation details of substreams and *requiring* advanced specs to depend on them, it will be difficult to change the compiler's substream logic in the future without breaking existing specs.

## The suggestion

Here's an alternative idea/concept/suggestion for how `pos` instances and substreams could look in KSY. I have no idea if this is in any way practical - I don't know a lot about how KSC works internally. This is just how in my mind these features would ideally work.

In short, I would completely hide the substream mechanics from the user, at least in the context of `pos` instances. Instead, I would specify all `pos` instances relative to an existing field, and let the compiler figure out how the IO should be performed (where to create a substream, etc.). For example:

```yaml
seq:
  - id: magic
    contents: "POTATOES"
  - id: header
    type: header
  - id: data_area
    size: header.len_data_area
  - id: trailer
    type: trailer
    size-eos: true
instances:
  thing_in_trailer:
    pos: trailer.ofs_thing
    relative-to: trailer
    type: thing
```

This would mean that the instance `thing_in_trailer` would start at `trailer.ofs_thing` bytes after the start of the `trailer` field. In this specific case, it's trivial to translate the code to the current substream-based syntax - `trailer` has an explicit size, so it has its own substream in exactly the place we need:

```yaml
instances:
  thing_in_trailer:
    pos: trailer.ofs_thing
    io: trailer._io
    type: thing
```

However, the advantage of the suggested syntax is that you can also specify fields in `relative-to` that don't have an exactly matching `_io`. For example, you could use it with non-explicitly-sized fields like `header` (which have an `_io`, but it doesn't match the start or size of the field), or with primitive fields like `data_area` (which don't have an `_io` at all).

Of course, this suggested syntax is more complex to implement in the compiler, especially for the cases where you don't already have a perfectly matching substream. Honestly though, I would already be happy if the compiler supported the syntax alone, even if it could only handle fields with a matching substream. It would already be a huge help if the compiler gave proper errors for `relative-to: a_primitive_field` and `relative-to: an_implicitly_sized_field`, rather than giving obscure errors about `_io` not existing, or silently doing the wrong thing (#563).

I hope my rant and suggestions made some sense :) If anything is unclear to you, please ask.

**Comment by Mingun** [comment]
I think we just need to change the starting point for `pos`. Now it goes from the [beginning of the stream|http://doc.kaitai.io/user_guide.html#_instances_data_beyond_the_sequence], and has to go from the beginning of the enclosing type (by default) or from beginning of the specified field (if `io` key specified). In such a case, it is better to replace the `io` key with the `origin` key at all, which will point to a specific field in `seq` of any type.

Then an example from [Choosing a substream](http://doc.kaitai.io/user_guide.html#_choosing_a_substream) section of guide could be rewritten as:
```yaml
seq:
  - id: global_header
    size: 1024
  - id: block_one
    type: big_container
    size: 4096
  - id: block_two
    type: smaller_container
    size: 1024
types:
  big_container:
    seq:
      - id: some_header
        size: 8
      # the rest of the data in this container would be referenced
      # from other blocks
  smaller_container:
    seq:
      - id: ofs_in_big
        type: u4
      - id: len_in_big
        type: u4
    instances:
      something_in_big:
        origin: _root.block_one
        pos: ofs_in_big
        size: len_in_big
```
which will work even if `block_two` will be defined, as
```yaml
  - id: block_two
    type: smaller_container
```
(without `size`)

**Comment by dgelessus** [comment]
@Mingun Right, I think that's basically what I'm suggesting.

After thinking about this some more, there are a couple of things that I didn't think about/explain properly in my original suggestion.

First, I wasn't very clear about what the maximum value of `pos` is when you use `relative-to: some_field`. With `io: some_field._io` this is obvious: `pos` can't point past the end of the stream. I'd suggest defining `relative-to` the same way - `pos` can only point *into* the field, not past the end. (The attribute could also be named differently to clarify this - perhaps `in: some_field`?)

Second, what kind of values does `relative-to` accept? It can't accept arbitrary expressions - the value needs to be something that has a real position in a stream, i. e. a field or another `pos` instance. There's also the question of how much we want to allow in terms of indirection, conditionals, etc. For example, with substreams it's perfectly valid to write `some_condition ? _parent.a_field._io : a_value_instance.a_field._io`, but supporting this level of complexity for `relative-to` would probably be difficult to implement in the compiler.

**Comment by KOLANICH** [comment]
>would start at trailer.ofs_thing bytes after the start of the trailer field.

After the end seems to be better.

I myself thought about something like that when coding https://github.com/KOLANICH/installer-framework/blob/patch-1/doc/qt_installer_framework.ksy
or
https://github.com/kaitai-io/kaitai_struct_formats/pull/264, though it was not about setting the `origin point` on each `instance`, but about setting it inside of the type that is instantiated. I also remember that I have even posted the ideas somewhere, but I cannot find them now.

IMHO `relative-to` has nothing to do with substreams. Substreams is mainly not about the offset (one can always manually compute the needed offset, so it is possible even there is no streams), but mostly is about total size of the substream (it allows to check bound and specify that some property must be strictly within bytes of some other property).

The issues about substreams mechanics you have mentioned are definitely bugs in KSC. IMHO everything must have a substream. But unstead of copying bytes the compiler must use memory views. And when unnneded substreams must be just optimized out.

**Comment by dgelessus** [comment]
> After the end seems to be better.

I don't think that would be better in any way. It would make it impossible to specify the start of a type as the base for the offset, and you no longer have an obvious restriction for range that `pos` can be in.

Also, I don't know how common this is, but I've seen multiple formats that contain "data areas", each of which contains one kind of structure at arbitrary locactions in the area. In that case it's very useful to say "this structure is inside this data area, at this offset", and have KS automatically check that `pos` is actually in range, and then calculate the correct offset. You can do the same thing manually of course (you can make a `value` instance for "start of data area X" and calculate all offsets based on that), but it's more work and doesn't do any range checking (not only can you point past the end of the area, you can also accidentally point *before* the area if you have a mistake in your calculation and end up with a negative number somehow).

An example of this that I'm currently working on is the [Mac resource fork format](https://github.com/dgelessus/resource_fork.ksy/blob/master/src/resource_fork.ksy). It has one area containing the data of all resources, one area for the names of all resources, etc. The resources themselves then refer to data within those areas, e. g. "name at offset 10 in the name area", "data at offset 500 in the data area". In that case it's very useful to write:

```yaml
instances:
  name:
    pos: ofs_name
    relative-to: name_area # ≈ io: name_area._io
    type: name
  data_block:
    pos: ofs_data_block
    relative-to: data_area # ≈ io: data_area._io
    type: data_block
```

If `relative-to` was based on the *end* of the field, I would have to specify whatever field comes *before* the name area, data area, etc., if such a field even exists. That would be much less readable in this case. (Again, I don't know how common this pattern is in other formats.)

> IMHO `relative-to` has nothing to do with substreams. Substreams is mainly not about the offset (one can always manually compute the needed offset, so it is possible even if there is no substreams), but mostly is about total size of the substream (it allows to check bound and specify that some property must be strictly within bytes of some other property).

What I'm saying is that substreams should (IMHO) be purely internal to the compiler. KS should of course have a way to say "this structure is somewhere in this range of data", but I think the mechanism for that should not be substreams (or any kind of IO streams at all). That's why I'm suggesting that `relative-to: some_field` should mean "this structure is somewhere in the range of data that `some_field` was parsed from".

> The issues about substreams mechanics you have mentioned are definitely bugs in KSC. IMHO everything must have a substream.

I think this is very similar to what I'm suggesting, except that you're calling them substreams and I'm calling them something-else-that-isn't-the-same-as-substreams :)

> But unstead of copying bytes the compiler must use memory views. And when unnneded substreams must be just optimized out.

This is exactly why I want to hide the concept of substreams from the KSY language, and instead introduce another mechanism for "ranges of data" that isn't dependent on concrete compiler/parser internals. Because advanced specs currently have to directly interact with IO streams and substreams, the compiler is limited in what optimizations it can do.

**Comment by KOLANICH** [comment]
>There's also no longer an obvious restriction for the range that pos can be in.

Of course it is - untill the end of the substream.

>In that case it's very useful to say "this structure is inside this data area, at this offset", and have KS automatically check that pos is actually in range, and then calculate the correct offset. 

Yes. Substreams are all about it. IMHO it should work as just defining a property and then using its `_io` as `io`. Or defining an own type for it in some cases, positioning instances inside of that type.

>In that case it's very useful to write:
>That's why I'm suggesting that relative-to: some_field should mean "this structure is somewhere in the range of data that some_field was parsed from".

Now I understand. Do you just propose to drop `._io` and rename `io` to `relative-to`? IMHO renaming would only make confusion. `relative-to` makes impression tuat only offset is meaningful. `substreams` is just the current terminology for "this structure is somewhere in the range of data that some_field was parsed from", but I hope not exactly that, but with respect to `process`.

>It would make it impossible to specify the start of a type as the base for the offset,
>If `relative-to` was based on the end of the field, I would have to specify whatever field comes before the name area, data area, etc., if such a field even exists. 

I was assumming that `relative-to` means not what I currently call a `substream`, but positioning an instance relative to that field in a parent substream, i.e. outside of that substream. I understood 
```yaml
relative-to: a
pos: b
```

as syntax sugar for

```yaml
pos: lea(a) + b
```
given #84 is implemented.

**Comment by dgelessus** [comment]
> IMHO renaming would only make confusion. `relative-to` makes impression that only offset is meaningful.

Yes, the name `relative-to` isn't great (see also https://github.com/kaitai-io/kaitai_struct/issues/709#issuecomment-593108619). For now you can consider it a temporary placeholder name, and I'm open to suggestions for better names.

> `substreams` is just the current terminology for "this structure is somewhere in the range of data that some_field was parsed from"

Yes, but "substream" is also the current terminology for "an IO stream that the compiler/parser creates internally when parsing fields with an explicit `size`". My point is that the two concepts should be decoupled - i. e. when a KSY file contains a field with an explicit `size`, the compiler shouldn't be required to also create a matching real IO stream for that field. (The compiler is still *allowed* to create a matching real IO stream like it does currently, but it should also be allowed to *not* create one if the specific field/structure doesn't require it.)

> Do you just propose to drop `._io` and rename `io` to `relative-to`?

Initially that would be the only change, yes. The main point of this change is to allow using `pos` instances without having to refer to real IO streams, for the reasons I've explained above.

The new syntax would also allow adding features that are impossible to implement with the current IO-based syntax. For example if `byte_array_field` is a primitive byte array (i. e. only has a `size` and no `type`), the KSY expression language doesn't allow accessing `byte_array_field._io`. This is not just because the compiler currently doesn't generate such a substream, but also because in the expression language `byte_array_field` is just a byte array, and byte arrays don't always have an `_io` (because you can also create byte arrays from literals). This means that `io: byte_array_field._io` is impossible to support in the compiler without major changes to the expression language.

On the other hand, `relative-to: byte_array_field` would not be a problem (at least not on the syntax level), because it doesn't require dealing with IO streams directly. The compiler might internally generate a real IO stream for `byte_array_field` and use that, but the user shouldn't have to worry about that (and especially shouldn't have to worry about what the IO stream is called, what object it's stored on, and how it needs to be accessed).

> but I hope it is not exactly that, but with respect to `process`.

I haven't used `process` much myself yet, so I'm not sure what the current behavior is for substreams and `process`. But yes, if I say `relative-to: a_field_with_process`, I would expect the instance to be based on the *processed* version of `a_field_with_process`. (If you need to work on the unprocessed data, you can move the `process` from the field into another `pos` instance.)

> `lea(a)`

I'm a bit confused what you mean with that. First, I assume it's a reference to the [x86 `lea` instruction](https://en.wikipedia.org/wiki/Load_Effective_Address)? It's been a while since I had to write x86 assembly, so I had to look up what exactly `lea` is used for again, but that doesn't really explain what `lea(a)` would do in KS. I assume it's supposed to calculate the offset of the field `a`, but it's not clear relative to what (the start of an IO stream? if so, which one?).

In any case, as I've said above, `relative-to` is just a placeholder name and doesn't mean much. It's relatively similar to the current `io` attribute. It's not *just* an offset calculation (but one aspect of it is that it changes the base for the `pos` offset, just like `io` currently does).

**Comment by KOLANICH** [comment]
>My point is that the two concepts should be decoupled - i. e. when a KSY file contains a field with an explicit size, the compiler shouldn't be required to also create a matching real IO stream for that field. (The compiler is still allowed to create a matching real IO stream like it does currently, but it should also be allowed to not create one if the specific field/structure doesn't require it.)

So why not to pretend that there are always substreams, leaving whether to actually create it up to the compiler?

>I'm a bit confused what you mean with that. First, I assume it's a reference to the x86 lea instruction?

Yes, and the mnemonic is itself an acronym meaning `load effective address`.

>I assume it's supposed to calculate the offset of the field a, but it's not clear relative to what (the start of an IO stream? if so, which one?).

Relative to the beginning of the current substream. I assumme that every field always has a substream (either a real one existing in the generated source or a virtual one existing only in KSC internal representation). So when we do
```
seq:
   - id: a
     type: b
types:
  b:
    seq:
      - id: c
        size: 1
```
first a substream is created for the root type, then for the field `a` (the compiler must create a substream for every property) within the substream for the root type, then for the field `c` within the substream for `a`, and in this case there is no need to c4eate real substreamsm so the generated code doesn't have them.
